### PR TITLE
New feature (Controller namespace, router prefix, sub domain)

### PR DESCRIPTION
**Pull request by sn01615**
Is this ok?
```php
    $router->ns('Admin')->prefix('/admin')->group(function (\Bramus\Router\Router $router) {
        $router->get('login', 'LoginController@login');
    });
    
    $router->prefix('/novel')->get('/nothing', function () {
        echo 'yes';
    });
    
    $router->ns('Admin')->prefix('/admin')->get('/login', 'LoginController@login1');
    $router->get('/login', 'LoginController@login2');
    
    $router->ns('Admin')->prefix('/admin')->group(function (\Bramus\Router\Router $router) {
        $router->get('login', 'LoginController@login3');
        $router->ns('Novel')->prefix('/novel')->group(function (\Bramus\Router\Router $router) {
            $router->get('info', 'NovelController@getNovelInfo');
        });
    });
```
Static method
```php
$router->get('/(\d+)', '\App\Controllers\User::showProfile');
```

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -13,7 +13,7 @@
         }
     ],
     "require": {
-        "php": ">=5.3.0"
+        "php": ">=5.4.0"
```
Comment (ovflowd) [inline composer.json]: I don't think is good making the Router require `5.4.0` php version on this point. It could break the installation of people that are using it. Does your changes really need php 5.4.0?

File: src/Bramus/Router/Router.php [hunk 2]
```diff
@@ -47,6 +47,20 @@ class Router
      */
     private $namespace = '';
 
+    /**
+     * @var static Parent Router ...
+     */
+    private $parent = '';
+
+    public function __construct($args = [])
```
Comment (ovflowd) [inline src/Bramus/Router/Router.php]: Missing document blocks for `__construct`. And please use `array()` instead of `[]`

File: src/Bramus/Router/Router.php [hunk 3]
```diff
@@ -47,6 +47,20 @@ class Router
      */
     private $namespace = '';
 
+    /**
+     * @var static Parent Router ...
+     */
+    private $parent = '';
```
Comment (ovflowd) [inline src/Bramus/Router/Router.php]: Why the parent Router would be a string? Why its default value is a string?

File: src/Bramus/Router/Router.php [hunk 4]
```diff
@@ -246,11 +270,52 @@ public function getRequestMethod()
...
+            }
+        }
+        return $this;
+    }
+
+    /**
+     * @param $namespace
+     * @return Router
+     */
+    public function ns($namespace)
```
Comment (ovflowd) [inline src/Bramus/Router/Router.php]: Missing more extended document blocks.

File: src/Bramus/Router/Router.php [hunk 5]
```diff
@@ -246,11 +270,52 @@ public function getRequestMethod()
...
+    public function group(callable $callback)
+    {
+        $callback($this);
+        if ($this->parent) {
+            $this->parent->mergeChild($this);
+        }
+        return $this;
+    }
+
+    public function mergeChild(Router $router)
```
Comment (ovflowd) [inline src/Bramus/Router/Router.php]: How will this impact performance?

File: src/Bramus/Router/Router.php [hunk 6]
```diff
@@ -246,11 +270,52 @@ public function getRequestMethod()
...
+    /**
+     * @param $namespace
+     * @return Router
+     */
+    public function ns($namespace)
+    {
+        return (new static(['parent' => $this]))->setNamespace($namespace)->setBaseRoute($this->baseRoute);
+    }
+
+    public function group(callable $callback)
```
Comment (ovflowd) [inline src/Bramus/Router/Router.php]: Missing document block 😅

File: src/Bramus/Router/Router.php [hunk 7]
```diff
@@ -443,4 +504,14 @@ public function setBasePath($serverBasePath)
     {
         $this->serverBasePath = $serverBasePath;
     }
+
+    /**
+     * @param string $baseRoute
+     * @return Router
+     */
+    public function setBaseRoute($baseRoute)
```
Comment (ovflowd) [inline src/Bramus/Router/Router.php]: Adjust your document block 👍

File: tests/RouterTest.php [hunk 8]
```diff
@@ -59,6 +59,87 @@ public function testUri()
             );
         }
 
+        public function testNs()
```
Comment (ovflowd) [inline tests/RouterTest.php]: If possible describe what these tests actually do 😄

**Comment by ovflowd** [comment]
Also kind of off-topic but I liked your profile photo. Kobayashi-san Chi no Maid Dragon is a really good anime.

File: tests/RouterTest.php [hunk 9]
```diff
@@ -10,28 +10,221 @@ public function notfound()
...
-            // Default SERVER_PROTOCOL method to HTTP/1.1
-            $_SERVER['SERVER_PROTOCOL'] = 'HTTP/1.1';
+            // Create Router instance
+            $router = new \Bramus\Router\Router();
+
+            $router2 = clone $router;
+
+            ob_start();
+            // Define routes
+            $router2->prefix('/novel')->get('/info', function () {
```
Comment (ovflowd) [inline tests/RouterTest.php]: Does named functions also are still working? eg.: `->get('/info', 'MyNamespace\MyController@method')`

Comment (ovflowd) [reply tests/RouterTest.php]: Would also something like this work?

```
function ($router2) {
    $router2->get('/info', function () {};

    ....
}($router2->prefix('/novel'));

Comment (ovflowd) [reply tests/RouterTest.php]: Maybe something like:
```
$router2->prefix('/novel')->group($router) {}
```

Would be cool, what you think?

**Comment by sn01615** [comment]
```
_1 concurrent requests_

Lumen (5.8.4) (Laravel Components 5.8.*)
ab -n 10000 http://127.0.0.1:8008/
Requests per second:    599.93 [#/sec] (mean)

router  b7a45e4
ab -n 10000 http://127.0.0.1/
Requests per second:    1269.14 [#/sec] (mean)


_10 concurrent requests_

Lumen (5.8.4) (Laravel Components 5.8.*)
ab -n 10000 -c 10 http://127.0.0.1:8008/
Requests per second:    2224.82 [#/sec] (mean)

router  b7a45e4
ab -n 10000 -c 10 http://127.0.0.1/admin/novel/info
Requests per second:    3520.71 [#/sec] (mean)
```

**Comment by ovflowd** [comment]
Great work! ✨

**Comment by sn01615** [comment]
Since my English is limited, if this can be merged, I hope someone can help improve the documentation.

**Comment by ovflowd** [comment]
Of course @sn01615 👍 

我想评论我也看
Kobayashi San的五月龙是一个很棒的系列

File: .php_cs.dist [hunk 10]
```diff
@@ -26,7 +26,7 @@ $finder = PhpCsFixer\Finder::create()
 return PhpCsFixer\Config::create()
     ->setRules([
         '@PSR2' => true,
-        'array_syntax' => [ 'syntax' => 'short' ],
+        'array_syntax' => [ 'syntax' => 'long' ],
```
Comment (bramus) [inline .php_cs.dist]: Could you revert this one back to short? It provides lots of unnecessary changes in the code, unrelated to the PR

File: .php_cs.dist [hunk 11]
```diff
@@ -47,7 +47,7 @@ return PhpCsFixer\Config::create()
         'phpdoc_to_comment' => true,
         'phpdoc_trim' => true,
         'single_quote' => true,
-        'ternary_to_null_coalescing' => true,
+        'ternary_to_null_coalescing' => false,
```
Comment (bramus) [inline .php_cs.dist]: Good catch! This will keep compatibility with older PHP versions that don't support the coalescing operator. 👍

File: src/Bramus/Router/Router.php [hunk 12]
```diff
@@ -373,7 +440,63 @@ private function handle($routes, $quitAfterRun = false)
...
+     * Set sub routes uri prefix.
+     * @param $prefix string A router prefix such as /admin
+     * @return $this
+     */
+    public function prefix($prefix)
+    {
+        $router = clone $this;
+        $router->setSubPrefix($prefix);
+
+        return $router;
```
Comment (bramus) [inline src/Bramus/Router/Router.php]: I see here (and also elsewhere in the code) that you always clone the router. Is there a specific reason for this? Couldn't it be done by directly returning `$this`?

For example:

```php
public function prefix($prefix)
{
    $this->setSubPrefix($prefix);
    return $this;    
}
```

File: .php_cs.dist (hunk 10 above)
Comment (sn01615) [reply .php_cs.dist]: `As of PHP 5.4 you can also use the short array syntax, which replaces array() with [].`
If php5.3 is not considered, it can be restored.

File: src/Bramus/Router/Router.php (hunk 12 above)
Comment (sn01615) [reply src/Bramus/Router/Router.php]: Clone can generate new contexts. This is mainly for coding convenience.

File: .php_cs.dist (hunk 10 above)
Comment (bramus) [reply .php_cs.dist]: Ah, good catch! We should keep it then, and perhaps only move on to newer (supported) PHP versions only with a new major release (2.x)

**Comment by kmvan** [comment]
Invokes static method maybe not a good idea.
Not enough OOP and Geek.

**Comment by bramus** [comment]
As part of a cleanup I'm closing this PR. Feel free to reopen a new one, taking the feedback provided above into account. Also keep in mind I don't want to over-complicate things with this router.
